### DISCUSSION TITLE
MQTT troubles...

### DISCUSSION THREAD

**Discussion by markbellkosel** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#225] [Q&A]
I seem to have everything working nicely other than MQTT. I click on the MQTT Config page, enter info, hit save, save confirmed, then I go to homepage and down at the bottom I still see MQTT: CONNECT_FAILED

**Comment by visualapproach** [comment]
Version? MQTT broker?

**Reply by markbellkosel** [reply]
Mosquito on a Pi4. 

Latest versions of Mosquitto and your software.

**Reply by visualapproach** [reply]
I think you have to look for the solution in your broker. Credentials requirements and such.

**Comment by chunkysteveo** [comment]
We need screenshots, config details, values etc. The MQTT on the device works perfect, so you have done something wrong, and we need more information to try to help you.

**Reply by markbellkosel** [reply]
Correction: MQTT is connected but Home Assistant won't discover it. I'm totally 100% new to Home Assistant so I'll have to sort that out. Unless someone would like to help me out.

**Reply by chunkysteveo** [reply]
Ahh yes. The device isn't auto discovered, tou have to manually add the sensors that you want/need. The new Master release has a new sample integration for HA... check out the Home Assistant folder!!

**Reply by visualapproach** [reply]
> Ahh yes. The device isn't auto discovered

Yet 😃

**Reply by chunkysteveo** [reply]
😍😍😍😍

**Reply by visualapproach** [reply]
Dev branch updated. Device is autodiscovered in HA. Check config/devices and lookup "layzspa". Then click add to Lovelace. Requires a configured mqtt broker in ha ofc.

**Reply by markbellkosel** [reply]
Damn I feel like an idiot for asking this but this means I need to run an ota update? Or do I need to do something with HA.

**Reply by visualapproach** [reply]
You need to update ESP with todays development version. 6-wire only for now. If you do it by usb or OTA is not important.
Regarding HA you have to comply with what I wrote in previous message.
It's fine to be newbie, just try to be specific in what you need help with. I won't write a complete manual here to cover all possible things you may or may not know. Not trying to be rude here, just helping you getting better help.
For example, "I use platformio and have a pump model xxx with 6 wires. How do I update over the air?" (If that's what you meant to ask 😀 )

**Reply by Skeletorjus** [reply]
> Dev branch updated. Device is autodiscovered in HA. Check config/devices and lookup "layzspa". Then click add to Lovelace. Requires a configured mqtt broker in ha ofc.

This is really amazing, and I simply can't wait to test it out!

I had a quick look through the code, and since the entities are set up by the device, the unique_id could include some device specific identifier to avoid any collision with other ID's (even if it's highly unlikely).  A common approach is to incorporate the MAC-address  in the ID. Unlike when setting unique_id in YAML, these are not user-facing, so they don't need to be pretty - just unique ;)

**Reply by Skeletorjus** [reply]
Just tried autodiscovery - working great!

**Comment by markbellkosel** [comment]
I totally understand. Wasn't taken as rude at all and I appologize for the vagueness.  

I do actually have a 6 wire "Miami" 2021 model. Since it's not the easiest to update over usb once assembled I do actually need help to update OTA. I followed everything you have published in the instructions (using platformio etc)

**Reply by markbellkosel** [reply]
I am also not entirely sure how to get the controls into a card. I have a very intricate home automation set up with Hubitat and it was recommended that I use Home Assistant if I want to get into making these kinds of things (esp based) and then use them in Hubitat via the bridge that was made.

**Reply by visualapproach** [reply]
> I totally understand. Wasn't taken as rude at all and I appologize for the vagueness.  
> 
> 
> 
> I do actually have a 6 wire "Miami" 2021 model. Since it's not the easiest to update over usb once assembled I do actually need help to update OTA. I followed everything you have published in the instructions (using platformio etc)

Good! Download the new files and open the folder/project in platformio. Find the file "platformio.ini" and edit it. At the end there is both some helping text and there is a line "something something esptool". Comment the line out by adding a ';' like the other lines above. Bonus tip: press ctrl+' to toggle comments. Remove or toggle the semi colons on the rows below. Enter the device's ip number on the appropriate line. There is a screenshot in the build instructions pdf iirc.
Save file (ctrl+s)
Click upload as usual (the right-pointing arrow at the bottom)
Wait for it to upload and restart. Will take a few minutes.
Done.
In home assistant left menu bar- go to configuration at the bottom. Click integrations, and choose devices from the top(bottom?) menu. Click on layzspa. Click 'add to Lovelace' on both cards. Done. You find the new cards in overview somewhere.

**Reply by markbellkosel** [reply]
Ok, I think I got it OTA updated correctly. I have to build before uploading correct? Everything seems functional. I went in to HA and am having trouble finding it. I think they made some changes in my most recent version. 

 Clicked Configuration

There is no "integrations" but there is "Devices & Services" I clicked that and it says Integrations, Devices, Entities, and Areas across the top.

I clicked devices and I see a list:

Forecast

Home Assistant Operating System

Mosquitto broker

Studio Code Server

**Comment by markbellkosel** [comment]
I think I may have other issues. When I check the web UI for the spa its going nuts. things are changing on it's own settings are reverting etc.

**Reply by markbellkosel** [reply]
It's like 18 outside here and I think stuff didn't agree. It's just strange that it started happening right after update

**Reply by visualapproach** [reply]
Did you change to 2021 model?

**Reply by markbellkosel** [reply]
Ahhh! I forgot about that part! Basically just redo what is in the instructions?

**Reply by visualapproach** [reply]
Yep

**Comment by markbellkosel** [comment]
Works great now! Now back to HA integration

**Comment by markbellkosel** [comment]
I still am not seeing it in HA

**Reply by visualapproach** [reply]
What have you done to expect it to be seen?

**Reply by markbellkosel** [reply]
I went in to my MQTT broker addon and clicked re configure with auto discover

**Reply by visualapproach** [reply]
Again, not to be sarcastic, but it's like saying that your car doesn't work, and when I ask about it you say you've popped the hood and expect me to know what is wrong.
You need to configure the mqtt broker to listen on a port, usually 1883. And make or have a user with credentials. Then enter all the settings in the esp mqtt config page and save.
If it works the device will be seen in HA where you described it in an earlier post.
Wish you luck

**Reply by markbellkosel** [reply]
Sorry. I already did all that. I verified that I have connected to the broker. I used MQTT explorer to verify. I just haven't seen the lazyspa show up in the HA ui

**Reply by visualapproach** [reply]
Restart ha

**Reply by markbellkosel** [reply]
Ok, if I go into "Entities" I see multiple stuff about spa listed but they all have a red "roll back" type icon and seem to be inoperable.

**Reply by visualapproach** [reply]
Go to devices, remove layzspa and restart ha then esp

**Reply by markbellkosel** [reply]
Lazyspa isn't even listed under devices at all. The entities are the only indication anywhere of spa related stuff. Should I remove those and restart?

**Reply by visualapproach** [reply]
Yes and make sure you have updated version of ha

**Comment by markbellkosel** [comment]
![Capture1](https://user-images.githubusercontent.com/64853248/158071405-63e58a33-3d9e-4f4e-9908-481b4f0118bb.JPG)

I am on 2022.3.4

I looked through Entities as aswell and they arent there anymore.

Is there something that could be wrong with configuration.yaml file?

**Reply by visualapproach** [reply]
Remove all old yaml related to layzspa. Restart

**Reply by markbellkosel** [reply]
unless I am looking in the wrong place, I don't see any.

![Capture](https://user-images.githubusercontent.com/64853248/158071739-cdc743fd-e929-4fb4-8aa0-2eadce7e2723.JPG)

**Reply by markbellkosel** [reply]
would it help to show configuration.yaml?

**Reply by markbellkosel** [reply]
![Capture](https://user-images.githubusercontent.com/64853248/158072326-c08f159c-334c-49ac-98a0-4b0bca22b417.JPG)

Just confirmed again that it is connected. I activated the bubbles toggle from UI and saw new messages come through in MQTT Explorer.

**Reply by markbellkosel** [reply]
I THINK I am missing stuff in configuration.yaml

![Capture](https://user-images.githubusercontent.com/64853248/158072667-014f0025-ff46-4974-9e07-a1b0220d3e6f.JPG)

**Reply by visualapproach** [reply]
Check the mqtt Explorer while restarting the esp. It should send a bunch of homeassistant/layzspa... topics at startup. Those are only sent at startup and is what makes ha recognize the device. You need auto discovery on in ha. Which I think is default but you can google it

**Reply by markbellkosel** [reply]
It is definitely on an is on by default. I will check the explorer and report back

**Reply by markbellkosel** [reply]
I don't see any mentions of homeasistant anywhere when I have Explorer pulled up on my laptop and then restart esp from web GUI on my phone

**Reply by visualapproach** [reply]
Sounds like you didn't installed the latest dev branch. Last function in main.cpp should be hasetup()

**Reply by markbellkosel** [reply]
yeah, it's not. hmmm I thought I did when I just click code and download zip
![Capture](https://user-images.githubusercontent.com/64853248/158080682-e17af458-3127-4e58-904b-b07c98fc2f48.JPG)

**Reply by markbellkosel** [reply]
oops it looks like I wasnt in the development branch

**Reply by markbellkosel** [reply]
Authentication needed now? Must be something different in development branch version? 
![Capture](https://user-images.githubusercontent.com/64853248/158081945-55ad70e5-377c-401d-a099-4b057ccb1c8f.JPG)

**Reply by markbellkosel** [reply]
SUCCESS!!!! I finally have it in HA! thank you!

**Comment by markbellkosel** [comment]
hmmmmm this is strange......doesn't feel that hot
![Capture](https://user-images.githubusercontent.com/64853248/158085547-229323ce-cd7e-42ce-8110-f81c379eb229.JPG)

**Reply by markbellkosel** [reply]
Odd behavior. If I make adjustments it show the typical 104 but its very strange that it then seems to physically turn the temp down and then also shows 200 plus

**Comment by markbellkosel** [comment]
I did some further investigation and it's even weirder.

I just did some maintenance and cleaning so the water temp is super low. Display says 50F but home assistant says 122. An offset of 72. However, I have it set to 104F and Home Assistant says it's set to 219. That's an offset of 115. Very odd! Any thoughts @visualapproach?

**Comment by markbellkosel** [comment]
Since we use Fahrenheit instead of Celsius I have the temperature unit switch turned off. I decided to try switching it to on. That made the GUI and pump unit display Celsius of course but that then made the device in HA display correctly in fahrenheit!

**Reply by visualapproach** [reply]
Hi @markbellkosel you are welcome to read chapter 8 in the updated build instructions, and update firmware to 2022-03-18 (DEV branch)
